clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainStorageEngine.java

```diff
@@ -0,0 +1,623 @@
+  }
+
+  @Override
+  public void destroy() {
+    storage.destroy();
+  }
+
+  @Override
+  public boolean shrink() {
+    return storage.shrink();
+  }
+
+  private static class DetachedChain implements Chain {
+
+    private final List<Element> elements;
+
+    private DetachedChain(List<Element> buffers) {
+      this.elements = unmodifiableList(buffers);
+    }
+
+    @Override
+    public Iterator<Element> reverseIterator() {
+      return Util.reverseIterator(elements);
+    }
+
+    @Override
+    public boolean isEmpty() {
+      return elements.isEmpty();
+    }
+
+    @Override
+    public Iterator<Element> iterator() {
+      return elements.iterator();
+    }
+
+  }
+
+  private static class PrimordialChain implements InternalChain {
+
+    private final ByteBuffer element;
+
+    public PrimordialChain(ByteBuffer element) {
+      this.element = element;
+    }
+
+    @Override
+    public Chain detach() {
+      throw new AssertionError("primordial chains cannot be detached");
+    }
+
+    @Override
+    public boolean append(ByteBuffer element) {
+      throw new AssertionError("primordial chains cannot be appended");
+    }
+
+    @Override
+    public boolean replace(Chain expected, Chain replacement) {
+      throw new AssertionError("primordial chains cannot be mutated");
+    }
+
+    @Override
+    public void close() {
+      //no-op
+    }
+  }
+
+  private final class AttachedInternalChain implements InternalChain {
+
+    /**
+     * Location of the chain structure, not of the first element.
+     */
+    private long chain;
+
+    AttachedInternalChain(long address) {
+      this.chain = address;
+      OffHeapChainStorageEngine.this.activeChains.add(this);
+    }
+
+    @Override
+    public Chain detach() {
+      List<Element> buffers = new ArrayList<Element>();
+
+      long element = chain + CHAIN_HEADER_SIZE;
+      do {
+        buffers.add(element(readElementBuffer(element), readElementSequenceNumber(element)));
+        element = storage.readLong(element + ELEMENT_HEADER_NEXT_OFFSET);
+      } while (element != chain);
+
+      return new DetachedChain(buffers);
+    }
+
+    @Override
+    public boolean append(ByteBuffer element) {
+      long newTail = createElement(element);
+      if (newTail < 0) {
+        return false;
+      } else {
+        long oldTail = storage.readLong(chain + CHAIN_HEADER_TAIL_OFFSET);
+        storage.writeLong(newTail + ELEMENT_HEADER_NEXT_OFFSET, chain);
+        try {
+          storage.writeLong(oldTail + ELEMENT_HEADER_NEXT_OFFSET, newTail);
+        } catch (NullPointerException e) {
+          throw e;
+        }
+        storage.writeLong(chain + CHAIN_HEADER_TAIL_OFFSET, newTail);
+        return true;
+      }
+    }
+
+    @Override
+    public boolean replace(Chain expected, Chain replacement) {
+      if (expected.isEmpty()) {
+        if (replacement.isEmpty()) {
+          return true;
+        } else {
+          insertHeader(replacement);
```

Comment:
We don't know (until we get inside the server) whether or not another operation has been appended to the chain since we pulled it that reinserts the mapping.